### PR TITLE
ENH: Integration of NMR data with units.

### PR DESCRIPTION
**Pull request by kfritzsc**
The `nmrglue.integration.integrate` and `nmrglue.integration.ndintegrate` functions should 
make it easier to integrate data within limits given in arbitrary units. This pull request is in response to Issue #44.

The `integrate` and `ndintegrate` fucntions can integrate data between limits given in any units that `fileiobase.unit_conversion` supports (this was done as suggested by @jjhelmus). In addition, if a noise range is defined the error of integration due to baseline noise is estimated. The `integrate` function can be used with 1D data, multiple integration ranges can be input to integrate multiple peaks within one spectrum.  The `ndintegrate` function can be used to integrate data of any dimensionality, however currently only one peak may be integrated at a time.

In response to @atomman comment in Issue #44 :

> ...I would prefer the unit_conv simply to be the ppm scale for the data...

A function to aid in the conversion of a ppm scale to a `unit_conversion` object was added as `fileiobase.uc_from_freqscale`. With input of a frequency scale in units of Hz, kHz or ppm and the observation frequency in Mhz `uc_from_freqscale` will create an `unit_conversion` object. This function does not work with time domain data arrays.

The addition of `uc_from_freqscale` also makes it easier to generate a `unit_conversion` object  from simulated testing data.

@jjhelmus I have added tests for each function but I am not sure that I put the tests in the right places. Should the tests go in the main directory folder (tests) or in the module's sub-directory?

### CODE REVIEW

File: tests/test_analysis_integration.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+def test_1d_integrate_withnoise():
+    """ Test integration of synthetic 1D data with two peaks and noise"""
+    # generate test scale
+    ppm_scale = _build_1d_ppm_scale()
+    uc = ng.fileio.fileiobase.uc_from_freqscale(ppm_scale, 100)
+
+    # generate test data
+    data = (multivariate_normal.pdf(ppm_scale, mean=5, cov=0.01) +
+            multivariate_normal.pdf(ppm_scale, mean=8, cov=0.01) * 2)
+    noise = np.random.randn(*data.shape) * 0.01
```
Comment (jjhelmus) [inline tests/test_analysis_integration.py]: It would probably be good to seed the random number generator to make this deterministic

**Comment by jjhelmus** [comment]
This looks great @kfritzsc.  Sorry for not getting to reviewing this earlier, I was traveling last week and let GtiHub emails pile up.  

The unit tests should go into the modules sub-directory, the tests in the main folder test directory are older tests which make use of the test data not included with nmrglue.  A long term goal is to remove these but I have not had the time to write up a good set of unit tests which do not depend on the test data to replace them.  I think I saw a style issue or two in the tests, if you wanted to check the file with pep8 when you move it I would appreciate it.

The integration.rst file should be added to the doc/source/reference/index.rst list under the analysis heading so that the new module get included in the documentation.  

Lastly it looks like the docstrings need a new additional spaces and odds and ends to properly render with Sphinx.  These are extra tricky to get correct, Sphinx and numpydocs are very unforgiving.  I'll add inline notes about the changes, if you want to skip these I can fix them before merging.

File: nmrglue/analysis/integration.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    .. math::
+        value = \sum_a^b s(x_{i}) dx
+
+    Where:
+    s is the signal, a and b are the limits of integration and dx is the width
+    of each bin.
+
+    The integration error due to baseline noise is calculated as:
+
+    ..math::
```
Comment (jjhelmus) [inline nmrglue/analysis/integration.py]: .. math::

A space between .. and math

File: nmrglue/analysis/integration.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        error = \sigma_{vol} = \sigma \sqrt{n},
+
+    if the noise_limits are set.
+
+    Where:
```
Comment (jjhelmus) [inline nmrglue/analysis/integration.py]: blank line needed before a math section

File: nmrglue/analysis/integration.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    .. math::
+        n = \frac{|b-a|}{dx}+1
```
Comment (jjhelmus) [inline nmrglue/analysis/integration.py]: n = \\frac{|b-a|}{dx}+1

Apparently \f is the line feed symbol and needs to be escaped.

File: nmrglue/analysis/integration.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        return values
+
+
+def ndintegrate(data, unit_conv, limits, unit='ppm', noise_limits=None):
+    """
+    Integrate one nD data array within limits given in units. Data points must
+    be equally spaced. Can only integrate one region per function call.
+
+    The integration error due to baseline noise is calculated as:
+    ..math::
```
Comment (jjhelmus) [inline nmrglue/analysis/integration.py]: A blank line is need before the math block and this should be
.. math::

File: nmrglue/analysis/integration.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        list of unit_conversion object associated with each dim of data.
+    limits: array like
+        With shape (2,) or (d, 2). 1D Array with lower and upper integration
+        limits for 1D . Or array with d rows of lower and upper integration
+        limits for each dimension.
+    noise_limits: Optional[array like]
+        With shape(2, ). Array with lower and upper limits to section of data
+        with only noise. A larger range will likely yield a more accurate
+        estimate. It is unwise to use the very end of the spectrum for the
+        noise range.
```
Comment (jjhelmus) [inline nmrglue/analysis/integration.py]: A blank line is needed before Returns

File: nmrglue/analysis/integration.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    Returns
+    -------
+    array
+        [value, ...] integration values
+
+    if noise_limits is given:
+
+    array
+        [[value, error], ...] where error a one sigma estimate of the error
+        only from the spectrum noise
```
Comment (jjhelmus) [inline nmrglue/analysis/integration.py]: Block comments should end with a blank line.

**Comment by kfritzsc** [comment]
@jjhelmus no need to apologize, this is an open source project! I appreciate all the time you do put into the project!

I was having some trouble seeing if the changes I made for the  sphinx problems fixed the issues. I was rebuilding the docs from scratch ( deleted the build directory and then using make html) but my changes were not reflected in the new html... I will be sure to figure this out before I submit any future code. Could you please fix any remaining sphinx related documentation errors this time?

**Comment by kfritzsc** [comment]
close #44

**Comment by jjhelmus** [comment]
Great work @kfritzsc, the Sphinx fixes looked fine.  Building the Sphinx docs locally is a bit challenging.  I'll be adding a PR shortly that should make this easier.

I added two commits 7f465a6ce and e9225b216a571 before merging to add the analysis/test directory to the package and fix a minor bug.  Let me know if either of this were incorrect.

**Pull request merged**